genes = ['N450', 'genome']

configfile: os.path.join(workflow.basedir, "defaults/config.yaml")

if os.path.exists("config.yaml"):
    configfile: "config.yaml"

include: "../shared/vendored/snakemake/config.smk"
include: "rules/config.smk"

rule all:
    input:
        auspice_json = expand("auspice/measles_{gene}.json", gene=genes),
        tip_frequencies_json = expand("auspice/measles_{gene}_tip-frequencies.json", gene=genes)

include: "rules/prepare_sequences.smk"
include: "rules/prepare_sequences_N450.smk"
include: "rules/construct_phylogeny.smk"
include: "rules/annotate_phylogeny.smk"
include: "rules/export.smk"

# Include custom rules defined in the config.
if "custom_rules" in config:
    for rule_file in config["custom_rules"]:

        include: os.path.join(os.getcwd(), rule_file)

rule clean:
    """Removing directories: {params}"""
    params:
        "results ",
        "auspice"
    shell:
        "rm -rfv {params}"
